### ISSUE TITLE
malformed json with --by-file-by-lang

### ISSUE THREAD

**Issue by bukzor** [AlDanial/cloc#308]
When I use cloc with options `--by-file-by-lang --json` the resulting json is malformed, like so: (In this case I'm looking at http://github.com/gothinkster/angularjs-realworld-example-app)

```
$ cloc --by-file-by-lang --vcs=git --json

{"header" : {
  "cloc_url"           : "github.com/AlDanial/cloc",
  "cloc_version"       : "1.76",
  "elapsed_seconds"    : 1.0764467716217,
  "n_files"            : 68,
  "n_lines"            : 2201,
  "files_per_second"   : 63.1707965434795,
  "lines_per_second"   : 2044.69004694409},
"src/js/services/user.service.js" :{
<snip>
  "language": "HTML"},
"src/js/profile/profile-articles.html" :{
  "blank": 0,
  "comment": 0,
  "code": 1,
  "language": "HTML"},
"SUM": {
  "blank": 470,
  "comment": 95,
  "code": 1636,
  "nFiles": 68} }
"JavaScript" :{
  "nFiles": 46,
  "blank": 334,
  "comment": 75,
  "code": 1074},
"HTML" :{
  "nFiles": 20,
  "blank": 125,
  "comment": 20,
  "code": 523},
"JSON" :{
  "nFiles": 1,
  "blank": 0,
  "comment": 0,
  "code": 25},
"Markdown" :{
  "nFiles": 1,
  "blank": 11,
  "comment": 0,
  "code": 14},
"SUM": {
  "blank": 470,
  "comment": 95,
  "code": 1636,
  "nFiles": 68} }
```

The problem is the first (of two) occurance of "} }", which ends the json object. This means json tools can't parse it:

```
$ cloc --by-file-by-lang --vcs=git --json  | jq -c .
{"header":{"cloc_url":"github.com/AlDanial/cloc","cloc_version":"1.76",
<snip>
ml":{"blank":0,"comment":0,"code":1,"language":"HTML"},"SUM":{"blank":470,"comment":95,"code":1636,"nFiles":68}}
"JavaScript"
parse error: Expected string key before ':' at line 355, column 14
```

**Comment by bukzor** [comment]
I forgot to say: Thanks for this tool! I find it quite useful.

---

The simplest solution is to add an open-brace after the problematic close-brace, like so:

```
  "language": "HTML"},
"SUM": {
  "blank": 470,
  "comment": 95,
  "code": 1636,
  "nFiles": 68} }
{
  "JavaScript" :{
  "nFiles": 46,
  "blank": 334,
  "comment": 75,
  "code": 1074},
"HTML" :{
```

It would probably be more consistent to put the whole header in the second json-object though. FWIW [jq](https://stedolan.github.io/jq/) and many tools like it operate on newline-delimited json, so I would probably change `--by-file-by-lang --json` to output something more like:

```
{"type":"header","cloc_url":"github.com/AlDanial/cloc","cloc_version":"1.76","elapsed_seconds":1.08324313163757,"n_files":68,"n_lines":2201,"files_per_second":62.7744575654057,"lines_per_second":2031.86148678615}
{"type":"file","blank":20,"comment":1,"code":84,"language":"JavaScript","file":"src/js/services/user.service.js"}
{"type":"file","blank":14,"comment":7,"code":65,"language":"JavaScript","file":"gulpfile.js"}
<snip>
{"type":"lang","nFiles":1,"blank":0,"comment":0,"code":25,"lang":"JSON"}
{"type":"lang","nFiles":1,"blank":11,"comment":0,"code":14,"lang":"Markdown"}
{"type":"SUM","blank":470,"comment":95,"code":1636,"nFiles":68}
```

In fact that's exactly what I've done, with this little script:

```
#!/bin/sh
glitch='} }'

exec sed "0,/$glitch/ s/$glitch/&\n{/" |
jq -sc '
  def unnest(index; type):
    .[index]
  | del(.header, .SUM)
  | keys_unsorted[] as $key
  | {type: type} + .[$key] + {(type): $key}
  ;

    {type: "header"} + .[0].header
  , unnest(0; "file")
  , unnest(1; "lang")
  , {type: "SUM"} + .[1].SUM
'
```

**Comment by AlDanial** [comment]
Thanks for reporting this issue.  JSON's a bear to get right; if only it would accept trailing comma's things would be so much easier.

I'll have to use a new switch (``--json2``?) to implement your suggested format since it changes the existing data schema and therefore would surely break existing applications.  Probably be a week or two out, but I'll get this in before the next release (which will also happen soon).

**Comment by AlDanial** [comment]
Give 1fc94cf a try and let me know your thoughts on the data structure.  While ``jq`` doesn't like it, the output seems legit to me.  I wrote this JSON checker to confirm correctness:

<pre>#!/usr/bin/env python
import json
import sys
def json_load(File):  # {{{1
    ok = True
    try:
        fh = open(File, 'r')
    except IOError:
        print('Cannot read %s, %s' % (File, sys.exc_info()[1]))
        return
    try:
        x = json.load(fh)
    except:
        ok = False
        print('JSON parse error for "%s": %s' % (File, sys.exc_info()[1]))
    fh.close()
    if ok:
        print('OK: %s' % (File))
# 1}}}
def main():  # {{{1
    if len(sys.argv) < 2:
        print("Usage:  %s <JSON file(s)>" % sys.argv[0])
        raise SystemExit
    for F in sys.argv[1:]:
        json_load(F)
# 1}}}
if __name__ == "__main__": main()
</pre>

**Comment by bukzor** [comment]
I will.
I find `jq` to be quite reliable though, so I expect that it's right. Could you show me the error and the corresponding bit of the json?

Just for fun: `python -m json.tool` does much of what your script does. Adding a bit of bash recovers the exact same behavior.
```
#!/bin/sh
for f in "$@"; do
  echo -n "$f: "
  if python -m 'json.tool' "$f" /dev/null; then
    echo OK
  fi
done
```

Demo
```
$ tail 2.json 3.json wat.json
==> 2.json <==
{}

==> 3.json <==
3

==> wat.json <==
wat

$ ./json-test.sh 2.json 3.json wat.json
2.json: OK
3.json: OK
wat.json: No JSON object could be decoded
```

**Comment by AlDanial** [comment]
``python -m json.tool`` that's a great tip, thanks.

If you run the latest code with two files to produce JSON output, the Python methods say the JSON file is fine but ``jq`` doesn't:
<pre>
> cloc --json --out x.json --by-file-by-lang hello.c qsort_2.c
> jq -e x.json 
jq: error: x/0 is not defined at <top-level>, line 1:
x.json
jq: 1 compile error
</pre>

**Comment by AlDanial** [comment]
Yes, the output from that looks much better.  Since ``jq`` agrees with Python's ``json`` module that the file is valid, I take it this half of the issue is resolved?

**Comment by AlDanial** [comment]
``--by-file-by-lang --json`` should work correctly in the just-released v 1.78.
